gravelsieve/init.lua

```diff
@@ -240,47 +310,34 @@ local function swap_node(pos, meta, start)
 end
 
 -- place ores to dst according to the calculated probability
-local function random_ore(inv, src)
-	local num
-	for ore, probability in pairs(gravelsieve.ore_probability) do
-		if math.random(probability) == 1 then
-			local item = ItemStack(ore)
-			if inv:room_for_item("dst", item) then
-				inv:add_item("dst", item)
-				return true     -- ore placed
+local function move_random_ore(inv, item)
+	local running_total = 0
+	local probabilities = gravelsieve.process_probabilities[item]
+	local chosen = math.random()
+	for ore, probability in pairs(probabilities) do
+		running_total = running_total + probability
+		if chosen < running_total then
+			local ore_item = ItemStack(ore)
+			if not inv:room_for_item("dst", ore_item) then
+				return false
 			end
+			inv:add_item("dst", ore_item)
+			return true
 		end
 	end
-	return false    -- gravel has to be moved
+	return false    -- Failure, this shouldn't really happen but might due to floating point errors
 end
 
-
-local function add_gravel_to_dst(meta, inv)
-	-- maintain a counter for gravel kind selection
-	local gravel_cnt = meta:get_int("gravel_cnt") + 1
-	meta:set_int("gravel_cnt", gravel_cnt)
-
-	if (gravel_cnt % 2) == 0 then  -- gravel or sieved gravel?
-		inv:add_item("dst", ItemStack("default:gravel"))        -- add to dest
-	else
-		inv:add_item("dst", ItemStack("gravelsieve:sieved_gravel")) -- add to dest
-	end
-end
-
-
 -- move gravel and ores to dst
-local function move_src2dst(meta, pos, inv, src, dst)
+local function move_src2dst(meta, pos, inv, item, dst)
+	local src = ItemStack(item)
 	if inv:room_for_item("dst", dst) and inv:contains_item("src", src) then
 		local res = swap_node(pos, meta, false)
 		if res then                                     -- time to move one item?
-			if src:get_name() == "default:gravel" then  -- will we find ore?
-				if not random_ore(inv, src) then        -- no ore found?
-					add_gravel_to_dst(meta, inv)
-				end
-			else
-				inv:add_item("dst", ItemStack("gravelsieve:sieved_gravel")) -- add to dest
+			local processed = move_random_ore(inv, item)
+			if processed then
+				inv:remove_item("src", src)
```

Comment:
This might not be the best way to handle it.
This effectively says: If i can't put the generated item into the output inventory, act as if the transaction never happened, but continue the process

We could use
```
if not processed then
    return false
end
```
This effectively says: If i can't put the generated item into the output inventory, stop the process and act as if blocked

The original logic said: If i can't put the generated item into the output inventory, put gravel into the inventory... which, if you think about it, it not always possible...